### PR TITLE
[INLONG-2106] DataProxy expose metric data using prometheus HttpServer.

### LINKED ISSUE #2106: [Feature] DataProxy expose metric data using prometheus HttpServer.
### Description

DataProxy expose metric data using prometheus HttpServer.

### Use case

_No response_

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by luchunliang**
### Title Name: [INLONG-2106][Inlong-DataProxy] DataProxy expose metric data using prometheus HttpServer.

Fixes #2106 

### Motivation

*Explain here the context, and why you're making that change. What is the problem you're trying to solve.*

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)
  - If a feature is not applicable for documentation, explain why?
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/prometheus/PrometheusMetricListener.java [hunk 1]
```diff
@@ -105,7 +140,114 @@ public void snapshot(String domain, List<MetricItemValue> itemValues) {
...
+     * collect
+     * 
+     * @return
+     */
+    @Override
+    public List<MetricFamilySamples> collect() {
+
+        // total
+        CounterMetricFamily totalCounter = new CounterMetricFamily(metricName + ".total", "help",
+                Arrays.asList("dimension"));
```
Comment (shink) [inline inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/prometheus/PrometheusMetricListener.java:183]: I think the help information should be more detailed.

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/prometheus/PrometheusMetricListener.java [hunk 2]
```diff
@@ -105,7 +140,114 @@ public void snapshot(String domain, List<MetricItemValue> itemValues) {
...
+        totalCounter.addMetric(Arrays.asList(M_SEND_FAIL_SIZE), metricItem.sendFailSize.get());
+        //
+        totalCounter.addMetric(Arrays.asList(M_SINK_DURATION), metricItem.sinkDuration.get());
+        totalCounter.addMetric(Arrays.asList(M_NODE_DURATION), metricItem.nodeDuration.get());
+        totalCounter.addMetric(Arrays.asList(M_WHOLE_DURATION), metricItem.wholeDuration.get());
+        List<MetricFamilySamples> mfs = new ArrayList<>();
+        mfs.add(totalCounter);
+
+        // id dimension
+        CounterMetricFamily idCounter = new CounterMetricFamily(metricName + ".id", "help", this.dimensionKeys);
```
Comment (shink) [inline inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/prometheus/PrometheusMetricListener.java:204]: ditto

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/prometheus/PrometheusMetricListener.java (hunk 1 above)
Comment (shink) [reply inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/prometheus/PrometheusMetricListener.java:183]: And the names of metrics are usually joind with underscores in Prometheus.

```suggestion
        CounterMetricFamily totalCounter = new CounterMetricFamily(metricName + "_total", "The total number of dataproxy metrcs.",
                Arrays.asList("dimension"));
```

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/prometheus/PrometheusMetricListener.java [hunk 3]
```diff
@@ -40,17 +56,28 @@
...
+
 /**
  * 
  * PrometheusMetricListener
  */
-public class PrometheusMetricListener implements MetricListener {
+public class PrometheusMetricListener extends Collector implements MetricListener {
 
-    public static final Logger LOG = LoggerFactory.getLogger(MetricRegister.class);
+    private static final Logger LOG = LoggerFactory.getLogger(MetricRegister.class);
```
Comment (shink) [inline inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/prometheus/PrometheusMetricListener.java:69]: Whether this logger should get from `PrometheusMetricListener.class`?
```suggestion
    private static final Logger LOG = LoggerFactory.getLogger(PrometheusMetricListener.class);
```

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/prometheus/PrometheusMetricListener.java (hunk 1 above)
Comment (luchunliang) [reply inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/prometheus/PrometheusMetricListener.java]: CounterMetricFamily idCounter = new CounterMetricFamily(metricName + "&group=id",
                "The metrics of inlong dataflow.", this.dimensionKeys);

File: inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/prometheus/PrometheusMetricListener.java (hunk 2 above)
Comment (luchunliang) [reply inlong-dataproxy/dataproxy-source/src/main/java/org/apache/inlong/dataproxy/metrics/prometheus/PrometheusMetricListener.java:204]: CounterMetricFamily idCounter = new CounterMetricFamily(metricName + "&group=id",
                "The metrics of inlong dataflow.", this.dimensionKeys);

**Comment by shink** [comment]
Dose the role of the method `snapshot()` is to regularly upload latest values of the metrics to custom monitor system? If so, does it work in this case?

When HTTPServer receives an HTTP request, the `collect()` method will collect metrics samples and push them to Prometheus. So, does calling the `snapshot()` method guarantee that the latest metrics values can be pushed to Prometheus?

It's just my question. Thank you.

**Comment by luchunliang** [comment]
> y question. T
After calling the snapshot() method to extract the latest metrics values, the metrics will go to zero.
 MetricObserver snapshot latest values(List<MetricItemValue>) of the metrics periodically from JMX MBean, and MetricObserver notify all MetricListener(audit system and more different monitor systems).
The latest values of the metrics is the data of a period(for example: a minute).
In opensource version, Prometheus HTTP Server provide the accumulated value after node start to work for Prometheus monitor system.

**Review by dockerzhang — APPROVED** [review]

**Review by shink — APPROVED** [review]
+1

**Review by gosonzhang — APPROVED** [review]
+1

**Pull request merged**